core/objects.go

```diff
@@ -605,16 +568,66 @@ func (jb *JSONBuffer) UnmarshalJSON(data []byte) (err error) {
 	return
 }
 
+// AcmeCertificateRequest ist the struct that we use for parsing
+// requests from clients
+type AcmeCertificateRequest struct {
+	CSR   *x509.CertificateRequest // The CSR
+	Bytes []byte                   // The original bytes of the CSR, for logging.
+}
+
+type rawAcmeCertificateRequest struct {
+	CSR JSONBuffer `json:"csr"`
+}
+
+// UnmarshalJSON provides an implementation for decoding CertificateRequest objects.
+func (cr *AcmeCertificateRequest) UnmarshalJSON(data []byte) error {
+	var raw rawAcmeCertificateRequest
+	if err := json.Unmarshal(data, &raw); err != nil {
+		return err
+	}
+
+	csr, err := x509.ParseCertificateRequest(raw.CSR)
+	if err != nil {
+		return err
+	}
+
+	cr.CSR = csr
+	cr.Bytes = raw.CSR
+	return nil
+}
+
+// MarshalJSON provides an implementation for encoding CertificateRequest objects.
+func (cr AcmeCertificateRequest) MarshalJSON() ([]byte, error) {
+	return json.Marshal(rawAcmeCertificateRequest{
+		CSR: JSONBuffer(cr.CSR.Raw),
+	})
+}
+
+// CertificateRequest objects are entirely internal to the server.
+// They represent a client's request for a certificate.  The ID from
+// the request is used as the final segment of the generic certificate
+// URI.  Requests to that URI return the most recent certificate
+// resulting from that request.
+type CertificateRequest struct {
```

Comment:
I could be sold on changing this to something like `CertificateContext`.  Too many requests around already.